gradoop-flink/src/main/java/org/gradoop/flink/model/impl/operators/matching/single/cypher/operators/filter/functions/FilterAndProjectTriple.java

```diff
@@ -0,0 +1,180 @@
+  /**
+   * Property keys used for value projection of the source vertex
+   */
+  private final List<String> sourceProjectionPropertyKeys;
+  /**
+   * Property keys used for value projection of the edge
+   */
+  private final List<String> edgeProjectionPropertyKeys;
+  /**
+   * Property keys used for value projection of the target vertex
+   */
+  private final List<String> targetProjectionPropertyKeys;
+  /**
+   * Meta data describing the vertex embedding used for filtering
+   */
+  private final EmbeddingMetaData filterMetaData;
+  /**
+   * Source vertex propertyKeys of the embedding used for filtering
+   */
+  private final List<String> sourceFilterPropertyKeys;
+  /**
+   * Edge propertyKeys of the embedding used for filtering
+   */
+  private final List<String> edgeFilterPropertyKeys;
+  /**
+   * Target vertex propertyKeys of the embedding used for filtering
+   */
+  private final List<String> targetFilterPropertyKeys;
+  /**
+   * Vertex match strategy
+   */
+  private final MatchStrategy vertexMatchStrategy;
+
+  /**
+   * New FilterAndProjectTriples
+   * @param sourceVariable the source variable
+   * @param edgeVariable edge variabe
+   * @param targetVariable target variable
+   * @param predicates filter predicates
+   * @param projectionPropertyKeys property keys used for projection
+   * @param vertexMatchStrategy vertex match strategy
+   */
+  public FilterAndProjectTriple(String sourceVariable, String edgeVariable, String targetVariable,
+    CNF predicates, Map<String, List<String>> projectionPropertyKeys,
+    MatchStrategy vertexMatchStrategy) {
+
+    this.predicates = predicates;
+    this.sourceVariable = sourceVariable;
+    this.targetVariable = targetVariable;
+    this.vertexMatchStrategy = vertexMatchStrategy;
+
+    this.sourceProjectionPropertyKeys =
+      projectionPropertyKeys.getOrDefault(sourceVariable, new ArrayList<>());
+    this.edgeProjectionPropertyKeys =
+      projectionPropertyKeys.getOrDefault(edgeVariable, new ArrayList<>());
+    this.targetProjectionPropertyKeys =
+      projectionPropertyKeys.getOrDefault(targetVariable, new ArrayList<>());
+
+    filterMetaData = createFilterMetaData(predicates, sourceVariable, edgeVariable, targetVariable);
+    sourceFilterPropertyKeys = filterMetaData.getPropertyKeys(sourceVariable);
+    edgeFilterPropertyKeys = filterMetaData.getPropertyKeys(edgeVariable);
+    targetFilterPropertyKeys = filterMetaData.getPropertyKeys(targetVariable);
+
+  }
+
+  @Override
+  public void flatMap(Triple triple, Collector<Embedding> out) throws Exception {
+
+    if (sourceVariable.equals(targetVariable)) {
+      // Check if source and target are expected to be the same
+      if (!triple.getSourceVertex().getId().equals(triple.getTargetVertex().getId())) {
+        return;
+      }
+    } else if (vertexMatchStrategy.equals(MatchStrategy.ISOMORPHISM) &&
+      triple.getSourceVertex().getId().equals(triple.getTargetVertex().getId())) {
+      return;
```

Comment:
instead of multiple `return` statements one could use a `boolean isValid = true` which is set to false instead of `return` and then you can use `if (isValid && filter(triple))`